### PR TITLE
(macOS) FocusZone: Handle Nested Focus Zones

### PR DESCRIPTION
**Pull request by Saadnajmi**
### Platforms Impacted
- [ ] iOS
- [x] macOS
- [ ] win32 (Office)
- [ ] windows
- [ ] android

### Description of changes

Let's say we have a FocusZone that contains many child FocusZones. An example might be a list of cards, where the list is a FocusZone, and each card is a FocusZone. In that scenario, if focus is on an element in an _inner_ FocusZone and we press Tab, we want Focus to move out of the parent, and onto the next element outside the parent FocusZone (aka, the next focusable element outside our list of cards).

As implemented right now, pressing tab would ignore that we might be a nested FocusZone, and put focus on the next card, which itself is inside the parent focus zone.  This is a little hard to describe, so see the demo videos.

### Verification

I added a test to the "Focus Zone edge cases" section of our FocusZone test page that should illustrate it well. 

Before:
Bad, pressing tab in Inner FocusZone 1 moved focus to Inner FocusZone2

https://user-images.githubusercontent.com/6722175/162550912-782bce7d-cd78-426c-88b9-02ea75825c45.mov

After:
Good, Pressing Tab and Shift+Tab while inside any Inner FocusZone moved Focus out of the parent FocusZone.

https://user-images.githubusercontent.com/6722175/162550909-9db7bead-b61f-4c10-9580-44a04d912873.mov

### Pull request checklist

This PR has considered (when applicable):
- [ ] Automated Tests
- [ ] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

### CODE REVIEW

File: packages/components/FocusZone/macos/RCTFocusZone.m [hunk 1]
```diff
@@ -362,34 +363,38 @@ - (NSView *)nextViewToFocusOutsideZone:(FocusZoneAction)action
...
-		NSView *element = [nextFocusZone defaultKeyView];
-		if (element != nil)
+		nextViewToFocus = [self previousValidKeyView];
+		while([nextViewToFocus isDescendantOf:parentFocusZone])
 		{
-			nextViewToFocus = element;
+			nextViewToFocus = [nextViewToFocus previousValidKeyView];
+		}
+		
+		// If the previous view is in a FocusZone, focus on its' defaultKeyView
```
Comment (harrieshin) [inline packages/components/FocusZone/macos/RCTFocusZone.m:392]: spell check its'

File: apps/fluent-tester/src/FluentTester/TestComponents/FocusZone/FocusZoneTest.tsx [hunk 2]
```diff
@@ -45,7 +43,28 @@ const EdgeCasesFocusZone: React.FunctionComponent = () => {
       <FocusZone>
         <SubheaderText>FocusZone with no elements</SubheaderText>
       </FocusZone>
-    </View>
+      <SubheaderText>Nested FocusZones</SubheaderText>
+      <Button>Outside Focus Zone</Button>
```
Comment (harrieshin) [inline apps/fluent-tester/src/FluentTester/TestComponents/FocusZone/FocusZoneTest.tsx:47]: super nit: can you add some padding around here please?

File: apps/fluent-tester/src/FluentTester/TestComponents/FocusZone/FocusZoneTest.tsx [hunk 3]
```diff
@@ -45,7 +43,28 @@ const EdgeCasesFocusZone: React.FunctionComponent = () => {
...
+          <Text>Inner FocusZone 2</Text>
+          <View style={focusZoneTestStyles.nestedFocusZoneStyle}>
+            <FocusZone>
+              <ListOfCheckboxes />
+            </FocusZone>
+          </View>
+          <Button>Inside Focus Zone</Button>
+        </FocusZone>
+      </View>
+      <Button>Outside Focus Zone</Button>
```
Comment (harrieshin) [inline apps/fluent-tester/src/FluentTester/TestComponents/FocusZone/FocusZoneTest.tsx:66]: super nit: also padding around the button as well for aesthetics and enough space for blue ring to show up?

File: packages/components/FocusZone/macos/RCTFocusZone.m [hunk 4]
```diff
@@ -147,9 +147,10 @@ static inline BOOL IsHorizontalNavigationWithinZoneAction(FocusZoneAction action
 	return action == FocusZoneActionRightArrow || action == FocusZoneActionLeftArrow;
 }
 
+/// Returns the first parent view of the given view that is a FocusZone.
 static RCTFocusZone *GetFocusZoneAncestor(NSView *view)
```
[the remark this answers is not in the document]
Comment (Saadnajmi) [reply packages/components/FocusZone/macos/RCTFocusZone.m:151]: That's correct. Would you prefer "GetFirstFocusZoneParent"?

File: packages/components/FocusZone/macos/RCTFocusZone.m [hunk 5]
```diff
@@ -362,34 +363,38 @@ - (NSView *)nextViewToFocusOutsideZone:(FocusZoneAction)action
...
+	RCTFocusZone *focusZoneAncestor = GetFocusZoneAncestor(parentFocusZone);
+	while (focusZoneAncestor != nil)
 	{
-		nextViewToFocus = [self previousValidKeyView];
+		parentFocusZone = focusZoneAncestor;
+		focusZoneAncestor = GetFocusZoneAncestor(parentFocusZone);
 	}
 
-	if ([nextViewToFocus isDescendantOf:self])
+	if (action == FocusZoneActionTab)  // Advance to next zone
```
[the remark this answers is not in the document]
Comment (Saadnajmi) [reply packages/components/FocusZone/macos/RCTFocusZone.m:376]: https://github.com/microsoft/fluentui-react-native/blob/master/docs/pages/Components/FocusZone.md

Probably should update what works on win32 and works on macOS as this is pretty old afailk.

File: packages/components/FocusZone/macos/RCTFocusZone.m [hunk 6]
```diff
@@ -362,34 +363,38 @@ - (NSView *)nextViewToFocusOutsideZone:(FocusZoneAction)action
...
-	else  // action == FocusZoneActionShiftTab per the conditional in keyDown
+	
+	// Find the first view outside the FocusZone (or any parent FocusZones) to place focus
+	RCTFocusZone *parentFocusZone = self;
+	RCTFocusZone *focusZoneAncestor = GetFocusZoneAncestor(parentFocusZone);
+	while (focusZoneAncestor != nil)
 	{
-		nextViewToFocus = [self previousValidKeyView];
+		parentFocusZone = focusZoneAncestor;
+		focusZoneAncestor = GetFocusZoneAncestor(parentFocusZone);
```
[the remark this answers is not in the document]
Comment (Saadnajmi) [reply packages/components/FocusZone/macos/RCTFocusZone.m:373]: I did it this way to be slightly more efficient. Line 369 calls it the first time so that the check in line 470 doesn't have to call it. Line 373 is the "looping" call. With a while loop, I figured I have to call it once, and then in a loop

**Review by harrieshin — APPROVED** [review]

File: packages/components/FocusZone/macos/RCTFocusZone.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (Saadnajmi) [reply packages/components/FocusZone/macos/RCTFocusZone.m:151]: Actually, I think there's no harm in changing this method to just return the topmost FocusZone in both cases it's used. I'll keep the naming and change the behavior.

**Comment by chiuam** [comment]
`if focus is on an element in an inner FocusZone and we press Tab, we want Focus to move out of the parent, and onto the next element outside the parent FocusZone (aka, the next focusable element outside our list of cards).`

Does this behavior align with win32? I thought it was by design that nested FocusZones are "tabbable".

File: packages/components/FocusZone/macos/RCTFocusZone.m (hunk 6 above)
[the remark this answers is not in the document]
Comment (Saadnajmi) [reply packages/components/FocusZone/macos/RCTFocusZone.m]: Removed this code so redundant

**Comment by Saadnajmi** [comment]
> `if focus is on an element in an inner FocusZone and we press Tab, we want Focus to move out of the parent, and onto the next element outside the parent FocusZone (aka, the next focusable element outside our list of cards).`
> 
> Does this behavior align with win32? I thought it was by design that nested FocusZones are "tabbable".

I worked on this support because a client reported a difference in behavior between win32 and macOS. They didn't expect nested focus zones to be tabbable. I also confirmed that nested focus zones is a valid scenario with CXE. To confirm the behavior is the same on windows, I can test out my new edge case test on windows and see if we have the same behavior.

**Comment by samuelfreiberg** [comment]
@Saadnajmi @harrieshin Nested FocusZones aren't supported for FURN. If a user uses nested focuszones, they should expect unwanted behavior. 

When I developed Win32's FocusZone, we talked a lot with Web about nested FocusZone's. They said nested FZs have been extremely difficult and time-consuming. A bunch of bugs kept coming in and the code was extremely complicated. 

I believe we shelved nested FZs after that discussion for the time being

**Comment by Saadnajmi** [comment]
Verified that win32 does indeed move focus out of the parent focus zone when you press tab. I also changed the test so that the inner focus zones are a horizontal row of buttons to make it a little closer to the actual scenario (each inner FocusZone handles horizontal keys, while the parent FocusZone handles vertical keys / switching between each FocusZone).
